### PR TITLE
feat: add command to retrieve snyk code autofix diffs

### PR DESCRIPTION
**Pull request by bastiandoetsch**
### Description

This PR provides a new command that enables retrieval of unified diffs for DeepCode AI Autofix suggestions by an IDE, so that the different suggestions can be reviewed by a user in the UI and one can be selected and executed.

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] README.md updated, if user-facing
- [x] License file updated, if new 3rd-party dependency is introduced

🚨After having merged, please update the CLI go.mod to pull in latest language server.

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: application/server/server.go [hunk 1]
```diff
@@ -267,6 +267,7 @@ func initializeHandler(srv *jrpc2.Server, c *config.Config) handler.Func {
 						snyk.GetActiveUserCommand,
 						snyk.CodeFixCommand,
 						snyk.CodeSubmitFixFeedback,
+						snyk.CodeFixDiffsCommand,
```
Comment (bastiandoetsch) [inline application/server/server.go:270]: This publishes the command availability to the client.

File: application/server/server_test.go [hunk 2]
```diff
@@ -1034,6 +1035,48 @@ func runSmokeTest(repo string, commit string, file1 string, file2 string, t *tes
...
+			Arguments: []any{uri.PathToUri(scanParams.FolderPath), uri.PathToUri(issue.FilePath), issue.Id},
+		})
+		assert.NoError(t, err)
+		var unifiedDiffs []code.AutofixUnifiedDiffSuggestion
+		err = call.UnmarshalResult(&unifiedDiffs)
+		assert.NoError(t, err)
+		assert.Greater(t, len(unifiedDiffs), 0)
+		// don't check for all issues, just the first
+		break
+	}
```
Comment (bastiandoetsch) [inline application/server/server_test.go:1078]: The smoke test now includes testing the auto-fix retrieval via command on mt-us.

File: domain/ide/command/code_fix_diffs.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	if !ok {
+		return nil, errors.New("Failed to parse issue id.")
+	}
+
+	issue := cmd.issueProvider.Issue(id)
+	if issue.ID == "" {
+		return nil, errors.New("Failed to find issue.")
+	}
+
+	suggestions, err := cmd.codeScanner.GetAutoFixDiffs(ctx, folderPath, relPath, issue)
```
Comment (bastiandoetsch) [inline domain/ide/command/code_fix_diffs.go:83]: This gets the unified diffs from the DeepCode AI API

File: domain/ide/workspace/folder.go [hunk 4]
```diff
@@ -497,3 +497,27 @@ func (f *Folder) sendScanResults(processedProduct product.Product, issuesByFile
...
+			}
+			if issueKey == key {
+				issue = i
+				return false
+			}
+		}
+		return true
+	})
+	return issue
+}
```
Comment (bastiandoetsch) [inline domain/ide/workspace/folder.go:523]: Search for an issue in a folder/workspace is by its unique 'key'.

File: domain/snyk/command.go [hunk 5]
```diff
@@ -43,6 +43,7 @@ const (
 	// Snyk Code specific commands
 	CodeFixCommand        = "snyk.code.fix"
 	CodeSubmitFixFeedback = "snyk.code.submitFixFeedback"
+	CodeFixDiffsCommand   = "snyk.code.fixDiffs"
```
Comment (bastiandoetsch) [inline domain/snyk/command.go:46]: register a new command identifier for the retrieval of the unified diffs

File: infrastructure/code/autofix.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/snyk/snyk-ls/application/config"
+	"github.com/snyk/snyk-ls/domain/snyk"
+)
+
+// AutofixUnifiedDiffSuggestion represents the diff between the original and the fixed source code.
+type AutofixUnifiedDiffSuggestion struct {
```
Comment (bastiandoetsch) [inline infrastructure/code/autofix.go:29]: This is returned by the command.

File: infrastructure/code/autofix.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+	var response AutofixResponse
+	response, err = s.RunAutofix(span.Context(), options)
+	if err != nil || response.Status == failed.message {
+		logger.Err(err).Msg("error getting autofix suggestions")
+		return unifiedDiffSuggestions, err
+	}
+
+	return response.toUnifiedDiffSuggestions(baseDir, options.filePath), err
+}
```
Comment (bastiandoetsch) [inline infrastructure/code/autofix.go:56]: wrapper for the autofix raw response

File: infrastructure/code/autofix.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+			if err != nil {
+				logger.Err(err).Msg("Error getting autofix suggestions")
+				return nil, err
+			}
+			if len(suggestions) > 0 {
+				return suggestions, nil
+			}
+		}
+	}
+}
```
Comment (bastiandoetsch) [inline infrastructure/code/autofix.go:105]: calls the API, polling it for autofix results.

File: infrastructure/code/convert.go [hunk 9]
```diff
@@ -495,3 +499,33 @@ func (s *AutofixResponse) toAutofixSuggestions(baseDir string, filePath string)
 
 	return fixSuggestions
 }
+
+func (s *AutofixResponse) toUnifiedDiffSuggestions(baseDir string, filePath string) []AutofixUnifiedDiffSuggestion {
```
Comment (bastiandoetsch) [inline infrastructure/code/convert.go:503]: Calculate a unified diff of the autofix response

**Review by cat2608 — APPROVED** [review]
Awesome! 🎉 

We said that we are confident that the Smoke Tests are covering enough scenarios for the new functions. Wonder if the _ticker sends a trigger every second to its channel_ logic could be stressed 🤔

**Pull request merged**